File: frontend/lib/api/sharing.ts
import { api } from './client'
import type {
  DoctorInteraction,
  DoctorOption,
  PatientGrant,
  PatientMedication,
} from './types'

export const sharingApi = {
  getDoctors: () => api.get<DoctorOption[]>('/sharing/doctors'),
  getGrants: () => api.get<PatientGrant[]>('/sharing/grants'),
  createGrant: (doctor_id: string) => api.post<PatientGrant>('/sharing/grants', { doctor_id }),
  revokeGrant: (grantId: string) => api.delete<PatientGrant>(`/sharing/grants/${grantId}`),
  getMedications: () => api.get<PatientMedication[]>('/sharing/medications'),
  createMedication: (data: Omit<PatientMedication, 'id' | 'patient_id' | 'created_at' | 'updated_at'>) =>
    api.post<PatientMedication>('/sharing/medications', data),
  updateMedication: (id: string, data: Omit<PatientMedication, 'id' | 'patient_id' | 'created_at' | 'updated_at'>) =>
    api.patch<PatientMedication>(`/sharing/medications/${id}`, data),
  deleteMedication: (id: string) => api.delete(`/sharing/medications/${id}`),
  getInteractions: () => api.get<DoctorInteraction[]>('/sharing/interactions'),
}

File: backend/app/api/v1/sharing.py
from fastapi import APIRouter, Response, status

from app.api.deps import DB, PatientUser
from app.schemas.patient_sharing import (
    DoctorOption,
    GrantCreate,
    InteractionResponse,
    MedicationCreate,
    MedicationResponse,
    PatientGrantResponse,
)
from app.services import patient_sharing as sharing_service

router = APIRouter()


@router.get("/doctors", response_model=list[DoctorOption])
def get_doctors(_patient: PatientUser, db: DB):
    return sharing_service.list_doctors(db)


@router.get("/grants", response_model=list[PatientGrantResponse])
def get_grants(patient: PatientUser, db: DB):
    return sharing_service.list_patient_grants(db, patient.id)


@router.post("/grants", response_model=PatientGrantResponse, status_code=status.HTTP_201_CREATED)
def create_grant(body: GrantCreate, patient: PatientUser, db: DB):
    return sharing_service.create_grant(db, patient.id, body.doctor_id)


@router.delete("/grants/{grant_id}", response_model=PatientGrantResponse)
def revoke_grant(grant_id: str, patient: PatientUser, db: DB):
    return sharing_service.revoke_grant(db, patient.id, grant_id)


@router.get("/medications", response_model=list[MedicationResponse])
def get_medications(patient: PatientUser, db: DB):
    return sharing_service.list_medications(db, patient.id)


@router.post("/medications", response_model=MedicationResponse, status_code=status.HTTP_201_CREATED)
def create_medication(body: MedicationCreate, patient: PatientUser, db: DB):
    return sharing_service.create_medication(db, patient.id, body)


@router.patch("/medications/{medication_id}", response_model=MedicationResponse)
def update_medication(medication_id: str, body: MedicationCreate, patient: PatientUser, db: DB):
    return sharing_service.update_medication(db, patient.id, medication_id, body)


@router.delete("/medications/{medication_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_medication(medication_id: str, patient: PatientUser, db: DB):
    sharing_service.delete_medication(db, patient.id, medication_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/interactions", response_model=list[InteractionResponse])
def get_interactions(patient: PatientUser, db: DB):
    return sharing_service.list_patient_interactions(db, patient.id)
